Define the commitment to an empty account delta as `EMPTY_WORD`

PhilippGackstatter opened the pull request:
Define the commitment to an empty account delta as `EMPTY_WORD`.

This assumes that `account_delta::compute_commitment` is only called when no more changes to the delta will be made. Otherwise, using `nonce_increment == 0` as an indicator for an empty delta is not good enough (see https://github.com/0xMiden/miden-base/issues/1522#issuecomment-3034831973).

Ideally, we'd have a concept of sealing the transaction, i.e. a point after which no more changes can be made. Before the transaction is sealed, calling `compute_commitment` would fail, so it could only be called afterward (most likely by an auth procedure after it has incremented the nonce).

Misc changes:
- Removed the TODO for adding a delta commitment test called from foreign accounts, because due to the above, this scenario might never actually happen.
- `P2ANY` notes are allowed to not contain any assets. This is useful for the newly added test where a note must be consumed to make the transaction non-empty, but no state change to the account should occur to keep the account delta empty. For reference, P2ID notes may also contain no assets. cc @igamigo 
- Renames `nonce_increment` to `nonce_delta` everywhere for consistency.

closes #1522

bobbinth commented on crates/miden-lib/asm/kernels/transaction/lib/memory.masm:
```diff
@@ -1223,26 +1223,26 @@ end
+#! - nonce_increment is the value by which the nonce changed.
+export.incr_account_delta_nonce_increment
+    exec.get_account_delta_nonce_increment add mem_store.ACCOUNT_DELTA_NONCE_PTR
 end
 
 #! Returns the delta of the account's nonce.
 #!
 #! Inputs:  []
-#! Outputs: [nonce_delta]
+#! Outputs: [nonce_increment]
 #!
 #! Where:
-#! - nonce_delta is the value by which the nonce changed.
-export.get_account_nonce_delta
+#! - nonce_increment is the value by which the nonce changed.
+export.get_account_delta_nonce_increment
     mem_load.ACCOUNT_DELTA_NONCE_PTR
 end
```
I know we've started calling it `nonce_increment`, but I wonder if we should go back to `nonce_delta` (the delta would still be defined as the difference between old an new nonce):

- It would be more consistent with other parts of the account delta.
- It would avoid awkward naming like `incr_account_delta_nonce_increment` and `get_account_delta_nonce_increment`.

bobbinth commented on crates/miden-lib/asm/kernels/transaction/lib/account_delta.masm:
```diff
@@ -61,6 +61,27 @@ export.compute_commitment
     exec.rpo::squeeze_digest
     # => [DELTA_COMMITMENT]
+
+    # if the account delta is empty (no storage and vault changes and a nonce increment of zero)
+    # then the commitment to it is defined as the empty word
+    padw
+    # => [EMPTY_WORD, DELTA_COMMITMENT]
+
+    # the epilogue will ensure that if the state has changed, then the
+    # nonce must be incremented.
+    # so if the nonce increment is zero, then the delta must be empty
+    # important caveat: this is only true if the delta sees no further changes after computing
+    # the commitment
+    exec.memory::get_account_delta_nonce_increment
+    # => [nonce_increment, EMPTY_WORD, DELTA_COMMITMENT]
+
+    eq.0
+    # => [is_delta_empty, EMPTY_WORD, DELTA_COMMITMENT]
+
+    # If is_delta_empty EMPTY_WORD remains.
+    # If !is_delta_empty DELTA_COMMITMENT remains.
+    cdropw
+    # => [EMPTY_WORD, DELTA_COMMITMENT]
```
As I mentioned in https://github.com/0xMiden/miden-base/issues/1522#issuecomment-3036970718, I think we should check the validity of delta before returning. Basically, if the nonce delta is `ZERO`, we should assert that the rest of the delta is empty.

This could be done by saving the 2nd `RATE` element after line 50 above into a local variable (it would be great if we had `copy_digest` in the standard library; let's create an issue for this) and then loading and comparing it to the `DELATA_COMMITMENT`.

bobbinth reviewed (commented):
Looks good! Thank you. I left some comments inline - the main one related to my comment in https://github.com/0xMiden/miden-base/issues/1522#issuecomment-3036970718.

PhilippGackstatter commented:
@bobbinth Thank you for your comments! I've addressed all of them. I renamed `nonce_increment` to `nonce_delta` in a separate commit in case you want to omit it while reviewing.

mmagician reviewed (approved):
LGTM ✅

bobbinth reviewed (approved):
Looks good! Thank you!
